### ISSUE TITLE
Cancellations cannot settle the computation asynchronously

### ISSUE THREAD

**Issue by semmel** [fluture-js/Fluture#216]
As I understand it if one wants to reflect cancellation in the outcome of a future task the cancel function returned from the computation may call `resolve` or `reject` on behalf of the not-cancelled computation. However, if this is done asynchronously it has no effect on the outcome.
Quick demo:
```javascript
var futureTask = Fluture.Future(function(fail, succeed){
	const timer = setTimeout(succeed, 10000, "world");
	return function(){ 
		setTimeout(fail, 1000, "aborted!");
		clearTimeout(timer);
	};
});
var cancel = futureTask.fork(console.warn.bind(console), console.log.bind(console));
cancel(); // nothing gets logged, the forked computation never settles
```
I find this behaviour with [fun-task](https://github.com/rpominov/fun-task) as well. I figure in *fun-task* it is some kind of preventing memory leaks if a task never settles after cancellation but may hang on to the `run` i.e. `fork`-handlers. IMO that clean-up is better hooked up with `resolve` and `reject`.

Anyway, that limitation prevents building *Future* or *fun-task* wrappers for some apis, e.g. [cancelling `fetch`](https://developer.mozilla.org/en-US/docs/Web/API/AbortController):
```javascript
var fetchJsonComputation = function(fail, succeed){
    const controller = new AbortController(); // currently just Edge and Firefox
    fetch('https://blackhole.webpagetest.org', {signal: controller.signal})
    .then(r => r.json())
    .then(succeed, fail);
    return controller.abort.bind(controller);
};
var fetchBlackholeTask = Fluture.Future(fetchJsonComputation);
// this will kick-off the ajax request
var cancel = fetchBlackholeTask.fork(console.warn.bind(console), console.log.bind(console));
// this cancels the ajax request just fine but the fork never settles with the expected DOMExecption
cancel();
```
Demo that with promises cancellation is propagated to their reject-path:
```javascript
var ctrl = new AbortController();
fetch('https://blackhole.webpagetest.org', {signal: ctrl.signal})
.then(r => r.json())
.then(console.log.bind(console), console.warn.bind(console, "Cancelled!"));
ctrl.abort(); // ~> Cancelled! DOMException
```

How do I settle a computation asynchronously in the cancel function?

**Comment by Avaq** [comment]
> As I understand it if one wants to reflect cancellation in the outcome of a future task the cancel function returned from the computation may call `resolve` or `reject` on behalf of the not-cancelled computation. 

You misunderstand. Cancelling a future means "unsubscribing" from its result. The consumer says: "Actually, I'm no longer interested in what you're producing". The cancellation signal travels through the chain, and any running Future can use it to clear memory or perform other optimisations knowing that the consumer is not interested in the result anymore.

> I find this behaviour with fun-task as well.

Me and @rpominov have collaborated with regards to cancellation semantics.

> that limitation prevents building *Future* or *fun-task* wrappers for some apis, e.g. cancelling `fetch`

```js
// this will kick-off the ajax request
var cancel = fetchBlackholeTask.fork(console.warn.bind(console), console.log.bind(console));
// this cancels the ajax request just fine but the fork never settles with the expected DOMExecption
cancel();
```

> the fork never settles with the expected DOMExecption

It doesn't need to. The consumer (the party that called `fork`, and is therefore also in charge of calling `cancel`) *already knows* that it cancelled the operation. The continuations are not called by design, because by calling `cancel` you've indicated that you are not interested in a continuation.

> How do I settle a computation asynchronously in the cancel function?

Cancellation happens synchronously (or at least with fire & forget). Therefore, nothing needs to settle afterwards. You can do whatever it is you wanted to do *straight away* after calling `cancel`.

----

Did this clear it up? Please feel free to ask questions if you have any.

**Comment by Avaq** [comment]
> with promises cancellation is propagated to their reject-path

To extend a bit on my previous answer, the reason Promises have to send a signal "back down" after receiving a cancellation signal is because they are *multicast*. One consumer can cancel a Promise, which may have many consumer who need to then know that it got cancelled.

Futures are *unicast*. Each computation only has a single consumer. When this consumer cancels a computation, they themselves are already aware of that.

**Comment by Avaq** [comment]
The fact that you can squeeze a reject or resolve in between the `cancel` call and the completion of the cancellation is actually a bit of an undesired behaviour.

I never considered that a user would call to one of the continuations during cancellation. It's not what you should do. ;)

**Comment by Avaq** [comment]
:point_up: I've opened a PR that removes the undesired behaviour. After it is merged, it should no longer be possible to send a rejection or resolution signal after calling `cancel`, so that should alleviate some of the confusion you had.

**Comment by semmel** [comment]
Thank you @Avaq for your kind explanations! Now it clears up a little :-)

Indeed my misunderstanding stems from mixing up *unicast* (Futures) and *multicast* (Promises).
But you can however make a Future *multicast* via `Future.cache()`, can't you? This is what I like about Promises: Even before their single job is complete you can distribute them freely into different parts of the application.

Distributing Futures that way would not be possible without repeating their job every time on `fork`.  As I understand `Future.cache()` solves that distribution problem, but now there is the problem which call to all the the `fork`-returned `cancel()` functions actually aborts the job? I can think of two strategies: 
- inclusive, where all consumers need to cancel or,
- exclusive, where the first call to a `cancel()` function aborts the job.

In both cancellation flavours, the consumer no longer has control over aborting the job. Wouldn't it be interesting for the consumers of a Cached Future to be notified of the cancellation?

**Comment by Avaq** [comment]
The strategy for cancelling a multicast (cached) Future is *inclusive*.

Actually, *cancel* may be a bit of a misnomer, and *unsubscribe* may have been a better term.

One consumer might unsubscribe from the multicast Future, and the continuations provided by this consumer will no longer be called. Only when *all* consumers unsubscribe from the multicast Future will the underlying computation receive the `cancel` signal.

The upstream cancel signal (the function returned from the `Future` constructor) really only serves to optimise for the situation where the result is no longer needed.

This philosophy arises from the acknowledgement that one cannot really *cancel* an asynchronous operation. Once it is triggered, we have to assume that the side-effects already took place.

**Comment by Avaq** [comment]
If you are interested in "the result of a cancellation" you are most likely doing something like gracefully closing a connection upon cancellation. In this case I would urge you to look at [`Future.hook()`](https://github.com/fluture-js/Fluture#hook) instead. It abstracts the pattern of acquiring a resource, consuming it and disposing it. Cancellation is more suitable for external interruptions, such as the user changing their mind.

**Comment by semmel** [comment]
I figure that *aborting* a job in a way which requires some action by the rest of the application should be triggered and carried out inside the computation. 
On the other hand the `cancel` function returned from the computation should silently *clean up* the running computation, e.g. cancel an *abort* timer.

**Comment by semmel** [comment]
```javascript
var cancelableAjaxFuture = Fluture.Future((fail, succeed) => {
	const ctrl = new AbortController();
	const abortTimer = setTimeout(ctrl.abort.bind(ctrl), 4000);
	fetch('https://blackhole.webpagetest.org', {signal: ctrl.signal})
	.then( response => response.json() )
	.then( succeed, fail);
	return () => clearTimeout(abortTimer);
});
var unsubscribe = cancelableAjaxFuture.fork(console.warn.bind(console, "failure"), console.log.bind(console, "done"));
// ~> "failure" DOMException  or
// nothing if unsubscribe is not called within 4 secs.
```
Works as intended now.

**Comment by Avaq** [comment]
Ah. I see now what you are trying to achieve. Thank you for bearing with me. I hope the following solution will be to your liking:

```js
//    loadJson :: String -> Future Error JsonValue
const loadJson = url => Fluture.Future((fail, succeed) => {
  const ctrl = new AbortController();
  fetch(url, {signal: ctrl.signal})
  .then( response => response.json() )
  .then( succeed, fail );
  return () => ctrl.abort();
})

//    fiveSecondTimeout :: Future Error a
const fiveSecondTimeout =
  Fluture.rejectAfter(5000, new Error('Operation timed out'));

//    eventualBlackhole :: Future Error JsonValue
const eventualBlackhole = Fluture.race(
  loadJson('https://blackhole.webpagetest.org'),
  fiveSecondTimeout
);

//main
const cancel = eventualBlackhole.fork(
  console.warn.bind(console, "failure"),
  console.log.bind(console, "done")
);

//potential interruption (for illustrative purposes)
document.addEventListener('keyup', e => e.keyCode === 27 && cancel(), {once: true})
```

Let me go over these bits one by one:

1. `loadJson`: I've moved the URL to be an argument, and removed the notion of the request timeout. All that remains is a general json loader which propagates the cancellation signal to the underlying machinery.
2. `fiveSecondTimeout`: This is simply a Future that rejects after five second with an error when it's forked. Of course the timeout is cleared if this Future were to be cancelled.
3. `eventualBlackhole`: This is where the magic happens. I've implemented the request timeout as a *race against the clock*, if you will. We're setting up the request to `blackhole.webpagetest.org` to *race* against the five second timeout. The resulting Future will reject or resolve with the first result it gets. So if the request were to succeed within five second, you get the resulting Json. If the request were to fail within five second, you get the resulting Error. If the request would not settle before five second, you'd get the `Operation timed out`-error.
    Now, the *real* nice thing is that Fluture realises that once it has a winner to the race, whatever result the loser is working on will no longer be needed, so it sends the cancellation signal for you. This means that if the request settles in time, the `fiveSecondTimeout` Future is cancelled, and it will clear its internal `setTimeout`. If the request takes too long, it will automatically be cancelled and propagate that signal to the `AbortController`, as we specified.
4. `potential interruption`: The `cancel` we got back from forking is actually the one produced by `race`, and what it does is simply cancel both racing Futures. So if there were to be an *external* cancellation signal, that signal would clear time timeout as well as cancel the request.

**Comment by Avaq** [comment]
Another nice thing about this approach is that you can control the timeout time and error message on a per-request basis, which in my opinion is the part of your code which has the knowledge to tailor the best error message:

```js
//A timeout with a very specific error message.
Fluture.race(
  loadJson('https://blackhole.webpagetest.org'),
  Fluture.rejectAfter(10000, new Error('Could not load the black hole in under ten seconds'))
);
```

**Comment by semmel** [comment]
> I hope the following solution will be to your liking

It indeed wraps the browser's [abortable fetch api](https://developers.google.com/web/updates/2017/09/abortable-fetch) very nicely!

>I've implemented the request timeout as a race against the clock, if you will.

It's interesting to note the similarity with how you would naively implement it with Promises and non-abortable jobs, e.g. fetch in Safari and Chrome; Just combine the promise with a timeout rejection using `Promise.race`:
```javascript
Promise.race(
	fetch(url), 
	new Promise((res, rej) => setTimeout(rej, 5000, new Error("timed out!"))
).then(res => res.json())
.then(....
// WARNING: Dont' use this code!
// The timeout rejection will fire regardless of fetch and will
// cause an unhandled exception.
```
It is nice that this *race-to-timeout* pattern works fine with Futures, with the additional benefit of the losers being "canceled". 👍 

>Another benefit of this approach is that you can control the timeout time and error message on a per-request basis, moving the creation of the error message to the part of your code which, in my opinion, has the knowledge to tailor the best error message

I agree 100%

In summary my problems of understanding originate in a **naming problem in the documentation of [`fork`](https://github.com/fluture-js/Fluture#fork)** :
>If you wish to cancel the computation, you may use the function returned by fork

As discussed above, you say it is really an *unsubscribe*:
- If you have a Cached Future (you may not know) the computation is aborted when *the last* fork-consumer has called that "cancel" function.
- To abort the `fetch` in your example wrapper above you don't use the "cancel" function returned by `Future.race([]).fork` at all. Instead it's wired up internally.

**Comment by Avaq** [comment]
> It is nice that this *race-to-timeout* pattern works fine with Futures

:bowtie: 

> In summary my problems of understanding originate in a naming problem in the documentation of `fork`

Cancellation (and its documentation) went through several stages before ending in master. Originally, its scope was bigger than just "unsubscribing". I should really revise and extend the documentation. I'll leave this issue open as a reminder to do so.
